src/rgw/rgw_bucket.cc

```diff
@@ -534,40 +535,42 @@ int rgw_remove_bucket(RGWRados *store, rgw_bucket& bucket, bool delete_children)
 
   RGWRados::Bucket target(store, info);
   RGWRados::Bucket::List list_op(&target);
-  CephContext *cct = store->ctx();
   int max = 1000;
-
   list_op.params.list_versions = true;
 
-  do {
-    objs.clear();
+  ret = list_op.list_objects(max, &objs, &common_prefixes, NULL);
+  if (ret < 0)
+    return ret;
 
-    ret = list_op.list_objects(max, &objs, &common_prefixes, NULL);
-    if (ret < 0)
-      return ret;
+  if (!objs.empty() && !delete_children) {
+    lderr(store->ctx()) << "ERROR: could not remove non-empty bucket " << bucket.name << dendl;
+    return -ENOTEMPTY;
+  }
 
-    if (!objs.empty() && !delete_children) {
-      lderr(store->ctx()) << "ERROR: could not remove non-empty bucket " << bucket.name << dendl;
-      return -ENOTEMPTY;
-    }
+  string prefix, delimiter;
+
+  ret = abort_bucket_multiparts(store, cct, info, prefix, delimiter);
+  if (ret < 0) {
+    return ret;
+  }
```

Comment:
cool, this covers the `ENOTEMPTY` case. but i'm still not sure why `abort_bucket_multiparts()` needs to happen before object removal now - some clarifying comments would help here

Reply:
Agree, It will not help with concurrent delete bucket and uploads either multipart or simple.
store->delete_bucket calls check_bucket_empty
but even this will not guarantee that no garbage (from running uploads) will left after bucket been deleted.

Reply:
right. this reminds me of some earlier discussions in https://github.com/ceph/ceph/pull/15630#pullrequestreview-44639570